bitcoin/src/psbt/mod.rs

```diff
@@ -71,87 +309,440 @@ impl Psbt {
+                        &psbt_input.witness_utxo,
+                        &psbt_input.non_witness_utxo,
+                    ) {
+                        (Some(witness_utxo), _) => Ok(witness_utxo),
+                        (None, Some(non_witness_utxo)) => {
+                            let vout = tx_input.previous_output.vout as usize;
+                            non_witness_utxo.output.get(vout).ok_or(Error::PsbtUtxoOutOfbounds)
+                        }
+                        (None, None) => Err(Error::MissingUtxo),
+                    },
+                )) as Box<dyn Iterator<Item = Result<&TxOut, Error>>>
             }
-        })
+            _ => {
+                // In PsbtV2, Input contains all the details, no need of unsigned_tx
+                Box::new(inner.inputs.iter().map(|input| {
+                    match (&input.witness_utxo, &input.non_witness_utxo) {
+                        (Some(witness_utxo), _) => Ok(witness_utxo),
+                        (None, Some(non_witness_utxo)) => {
+                            let output_index = input.output_index.unwrap() as usize;
+                            non_witness_utxo
+                                .output
+                                .get(output_index)
+                                .ok_or(Error::PsbtUtxoOutOfbounds)
+                        }
+                        (None, None) => Err(Error::MissingUtxo),
+                    }
+                })) as Box<dyn Iterator<Item = Result<&TxOut, Error>>>
+            }
+        }
     }
 
-    /// Checks that unsigned transaction does not have scriptSig's or witness data.
-    fn unsigned_tx_checks(&self) -> Result<(), Error> {
-        for txin in &self.unsigned_tx.input {
-            if !txin.script_sig.is_empty() {
-                return Err(Error::UnsignedTxHasScriptSigs);
+    /// Calculates the max lock time and assigns it to the given locktime variable.
+    fn max_locktime(locktime: &mut Option<absolute::LockTime>, new_lock: absolute::LockTime) {
+        match *locktime {
+            None => {
+                *locktime = Some(new_lock);
+            }
+            Some(lock) =>
+                if lock >= new_lock {
+                    *locktime = Some(lock);
+                } else {
+                    *locktime = Some(new_lock);
+                },
+        }
+    }
+
+    /// Adds a new [`Input`] to this PsbtV2.
+    ///
+    /// Only to be used for PsbtV2. Returns [`Error`] otherwise.
+    pub fn add_input(&mut self, input: Input) -> Result<(), Error> {
+        if self.inner.version == Version::PsbtV0 {
+            return Err(Error::Version("new inputs can not be added to PsbtV0"));
+        }
+
+        if self.inner.tx_modifiable.is_none()
+            || !self.inner.tx_modifiable.as_ref().unwrap().input_modifiable
+        {
+            return Err(Error::InputNotAddable("input modifiable flag is false"));
+        }
+        input.validate_version(self.inner.version)?;
```

Comment:
> Psbt maintains the invariant that PsbtInner is valid for the version, right?

This function mutates the Psbt, so the new given `Input` needs to comply with the version of the Psbt. Since `Input` is a public struct, it is possible to create a PsbtV0 Input (without the `output_index` and `previous_tx_id`) and then try to add it to a Psbt version 2. This check is only to maintain the guarantee.

Reply:
Derp, I have no idea what I was thinking yesterday, I read this as `self.validate` which it obviously is not. Thanks for you patient response.